### PR TITLE
fix: deno-types directive should have higher precedence than X-TypeScript-Types header

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes https://github.com/denoland/deno/issues/6749

### CODE REVIEW

File: cli/tests/type_headers_deno_types.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Following import uses two distinct ways to provide types:
+ * - X-TypeScript-Types headers
+ * - @deno-types directive
+ * 
+ * Because "@deno-types" directive must be placed by user explicitly it
+ * should have higher precedence than type header.
```
Comment (ry) [inline cli/tests/type_headers_deno_types.ts:7]: Here you write that `@deno-types` should have higher precedence, but in the title you say `X-TypeScript-Types` should have higher precedence. Which is it?

Comment (bartlomieju) [reply cli/tests/type_headers_deno_types.ts:7]: `@deno-types` should have higher precedence

**Comment by bartlomieju** [comment]
> I realize you're just fixing a bug and that's good. But I have to say that all this X-TypeScript-Type and `@deno-types` stuff is ugly and complex. It's starting to feel like `node_modules` resolution all over again. We need to find a way to simplify this.

I agree with you wholeheartedly

**Comment by kitsonk** [comment]
We have been around on this a few times.  I have yet heard a viable alternative approach to allowing people to use type definitions with plain old JavaScript.

The specific use cases exemplify the situation.  The types provided by the package and served by the CDN are incompatible with Deno, so the user is substituting types that are compatible with Deno at the import point.

We need two solutions to fit the two use cases of: "As a provider of a JavaScript module, I want to provide you types to use when type checking the consumption of this module" and the "I as an importer of a JavaScript module, know the types I want to use to type check that module".  The first one needs to be coupled to the import, the second one needs to impact the import site.  We have `X-TypeScript-Types` and trip-slash reference in JS for the first one and `@deno-types` for the second one.

I have yet to hear an alternative that meets the needs of the use cases that is an improvement on the current semantics.  The other option is to simply say that JavaScript imported into a Deno application cannot use external type definitions and gets typed as `any`.

**Comment by ry** [comment]
@kitsonk Yea I understand the need. It's just a relatively large complication for a relatively small problem. I don't have an alternative solution tho. I'm tempted to say - let's just not support supplying external type definitions to JS - but I suppose that probably precludes a lot of important situations (like React?).

**Comment by kitsonk** [comment]
Well, and a lot of folks who want to support Deno and Node.js with a single distributable of JavaScript plus TypeScript types.  It would mean a lot of folks would not be able to consume that JavaScript in a type safe way as the authors intended.

**Pull request merged**